File: java/src/main/java/com/kcl/api/Service.java
package com.kcl.api;

import com.kcl.api.Spec.*;

public interface Service {
    // Parse KCL single file AST JSON string
    ParseFile_Result parseFile(ParseFile_Args args) throws Exception;

    // Parse KCL program AST JSON string
    ParseProgram_Result parseProgram(ParseProgram_Args args) throws Exception;

    // Loads KCL package and returns the AST, symbol, type, definition information.
    LoadPackage_Result loadPackage(LoadPackage_Args args) throws Exception;

    // Loads KCL package and returns the AST, symbol, type, definition information.
    LoadPackage_Result loadPackageWithCache(LoadPackage_Args args) throws Exception;

    // Execute KCL file with args
    ExecProgram_Result execProgram(ExecProgram_Args args) throws Exception;

    // Override KCL file with args
    OverrideFile_Result overrideFile(OverrideFile_Args args) throws Exception;

    // List all the variables in the KCL file
    ListVariables_Result listVariables(ListVariables_Args args) throws Exception;

    // Service for getting the full schema type list
    GetSchemaType_Result getFullSchemaType(GetFullSchemaType_Args args) throws Exception;

    // Service for formatting a code source
    FormatCode_Result formatCode(FormatCode_Args args) throws Exception;

    // Service for formatting KCL file or directory path
    FormatPath_Result formatPath(FormatPath_Args args) throws Exception;

    // Service for KCL Lint API
    LintPath_Result lintPath(LintPath_Args args) throws Exception;

    // Service for validating the data string using the schema code string
    ValidateCode_Result validateCode(ValidateCode_Args args) throws Exception;

    // Service for building setting file config from args
    LoadSettingsFiles_Result loadSettingsFiles(LoadSettingsFiles_Args args) throws Exception;

    // Service for renaming all the occurrences of the target symbol in the files
    Rename_Result rename(Rename_Args args) throws Exception;

    // Service for renaming all the occurrences of the target symbol and rename them
    RenameCode_Result renameCode(RenameCode_Args args) throws Exception;

    // Service for the testing tool
    Test_Result test(Test_Args args) throws Exception;
}

File: swift/Sources/KclLib/Service.swift
import Foundation

// Define the protocol for services that manage various operations on KCL programs.
public protocol Service {
  // Parses a single KCL file and returns its Abstract Syntax Tree (AST) as a JSON string.
  func parseFile(_ args: ParseFileArgs) throws -> ParseFileResult

  // Parses a KCL program and returns the Abstract Syntax Tree (AST) in JSON format.
  func parseProgram(_ args: ParseProgramArgs) throws -> ParseProgramResult

  // Loads a KCL package and retrieves AST, symbol, type, and definition information.
  func loadPackage(_ args: LoadPackageArgs) throws -> LoadPackageResult

  // Executes a KCL file with provided arguments.
  func execProgram(_ args: ExecProgramArgs) throws -> ExecProgramResult

  // Overrides specified elements in a KCL file according to given arguments.
  func overrideFile(_ args: OverrideFileArgs) throws -> OverrideFileResult

  // Lists all variables declared in a KCL file.
  func listVariables(_ args: ListVariablesArgs) throws -> ListVariablesResult

  // Lists all options defined in a KCL program.
  func listOptions(_ args: ParseProgramArgs) throws -> ListOptionsResult

  // Retrieves the full schema type mapping for a KCL program.
  func getSchemaTypeMapping(_ args: GetSchemaTypeMappingArgs) throws -> GetSchemaTypeMappingResult

  // Formats source code according to KCL style guidelines.
  func formatCode(_ args: FormatCodeArgs) throws -> FormatCodeResult

  // Formats KCL files or directories to conform to style guidelines.
  func formatPath(_ args: FormatPathArgs) throws -> FormatPathResult

  // Runs linting checks on KCL files and reports errors and warnings.
  func lintPath(_ args: LintPathArgs) throws -> LintPathResult

  // Validates a data string against a schema defined in a KCL code string.
  func validateCode(_ args: ValidateCodeArgs) throws -> ValidateCodeResult

  // Builds configuration from settings files.
  func loadSettingsFiles(_ args: LoadSettingsFilesArgs) throws -> LoadSettingsFilesResult

  // Renames symbols across files within a KCL package.
  func rename(_ args: RenameArgs) throws -> RenameResult

  // Renames symbols in source code without modifying files directly.
  func renameCode(_ args: RenameCodeArgs) throws -> RenameCodeResult

  // Executes tests on KCL packages using specified test arguments.
  func test(_ args: TestArgs) throws -> TestResult

  // Updates dependencies for a KCL project based on defined specifications.
  func updateDependencies(_ args: UpdateDependenciesArgs) throws -> UpdateDependenciesResult

  // Retrieves version information about the KCL service.
  func getVersion(_ args: GetVersionArgs) throws -> GetVersionResult
}
